File: ios/vane_flutter/Sources/vane_flutter/VaneFlutterPlugin.swift
import Flutter
import UIKit

#if canImport(VaneSwift)
import VaneSwift
#endif

public class VaneFlutterPlugin: NSObject, FlutterPlugin {
  private var nextHandle: Int64 = 1
  private var clients: [Int64: VaneClient] = [:]
  private let lock = NSLock()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "vane_flutter", binaryMessenger: registrar.messenger())
    let instance = VaneFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "createClient":
      createClient(call, result: result)
    case "execute":
      execute(call, result: result)
    case "closeClient":
      closeClient(call, result: result)
    case "setCertificatePins":
      setCertificatePins(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func createClient(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      let args = try requiredDictionary(call.arguments)
      let config = try configFromMap(try requiredDictionary(args["configuration"]))
      let client = try createVaneClient(config: config)
      lock.lock()
      let handle = nextHandle
      nextHandle += 1
      clients[handle] = client
      lock.unlock()
      result(handle)
    } catch {
      result(FlutterError(code: "VANE_CREATE_CLIENT", message: "\(error)", details: nil))
    }
  }

  private func execute(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      let args = try requiredDictionary(call.arguments)
      let handle = try requiredInt64(args["handle"])
      let request = try requestFromMap(try requiredDictionary(args["request"]))
      guard let client = client(for: handle) else {
        result(FlutterError(code: "VANE_INVALID_CLIENT", message: "No Vane client exists for handle \(handle)", details: nil))
        return
      }

      Task {
        do {
          let response = try await client.execute(request)
          result(response.toMap())
        } catch {
          result(FlutterError(code: "VANE_EXECUTE", message: "\(error)", details: nil))
        }
      }
    } catch {
      result(FlutterError(code: "VANE_EXECUTE", message: "\(error)", details: nil))
    }
  }

  private func closeClient(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      let args = try requiredDictionary(call.arguments)
      let handle = try requiredInt64(args["handle"])
      lock.lock()
      clients.removeValue(forKey: handle)
      lock.unlock()
      result(nil)
    } catch {
      result(FlutterError(code: "VANE_CLOSE_CLIENT", message: "\(error)", details: nil))
    }
  }

  private func setCertificatePins(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      let args = try requiredDictionary(call.arguments)
      let handle = try requiredInt64(args["handle"])
      let host = try requiredString(args["host"])
      let pins = stringList(args["pins"])
      guard let client = client(for: handle) else {
        result(FlutterError(code: "VANE_INVALID_CLIENT", message: "No Vane client exists for handle \(handle)", details: nil))
        return
      }
      try client.setCertificatePins(host: host, pins: pins)
      result(nil)
    } catch {
      result(FlutterError(code: "VANE_SET_CERTIFICATE_PINS", message: "\(error)", details: nil))
    }
  }

  private func client(for handle: Int64) -> VaneClient? {
    lock.lock()
    defer { lock.unlock() }
    return clients[handle]
  }

  private func configFromMap(_ map: [String: Any]) throws -> VaneClientConfig {
    var config = createDefaultConfig()
    config.baseUrl = map["baseUrl"] as? String
    config.defaultHeaders = stringMap(map["defaultHeaders"])
    config.dnsOverrides = stringMap(map["dnsOverrides"])
    config.certificatePins = stringListMap(map["certificatePins"])
    config.cookiesEnabled = boolValue(map["cookiesEnabled"], fallback: config.cookiesEnabled)
    config.cookiePersistencePath = map["cookiePersistencePath"] as? String
    config.connectionPoolEnabled = boolValue(map["connectionPoolEnabled"], fallback: config.connectionPoolEnabled)
    config.maxIdleConnections = uint64Value(map["maxIdleConnections"], fallback: config.maxIdleConnections)
    config.connectionIdleTimeoutSeconds = uint64Value(
      map["connectionIdleTimeoutSeconds"],
      fallback: config.connectionIdleTimeoutSeconds
    )
    config.retryMaxAttempts = uint64Value(map["retryMaxAttempts"], fallback: config.retryMaxAttempts)
    config.retryInitialDelayMillis = uint64Value(
      map["retryInitialDelayMillis"],
      fallback: config.retryInitialDelayMillis
    )
    config.retryMaxDelayMillis = uint64Value(map["retryMaxDelayMillis"], fallback: config.retryMaxDelayMillis)
    config.retryUnsafeMethods = boolValue(map["retryUnsafeMethods"], fallback: config.retryUnsafeMethods)
    config.maxRequestBodyBytes = uint64Value(map["maxRequestBodyBytes"], fallback: config.maxRequestBodyBytes)
    config.maxResponseBodyBytes = uint64Value(map["maxResponseBodyBytes"], fallback: config.maxResponseBodyBytes)
    config.timeoutSeconds = optionalUInt64Value(map["timeoutSeconds"])
    config.followRedirects = boolValue(map["followRedirects"], fallback: config.followRedirects)
    config.userAgent = map["userAgent"] as? String
    config.protocolMode = protocolMode(map["protocolMode"] as? String)
    config.proxyUrl = map["proxyUrl"] as? String
    config.proxyAuthorization = map["proxyAuthorization"] as? String
    return config
  }

  private func requestFromMap(_ map: [String: Any]) throws -> VaneRequest {
    return VaneRequest(
      url: try requiredString(map["url"]),
      method: map["method"] as? String ?? "GET",
      headers: stringMap(map["headers"]),
      queryParams: stringMap(map["queryParams"]),
      body: map["body"] as? FlutterStandardTypedData != nil
        ? (map["body"] as! FlutterStandardTypedData).data
        : map["body"] as? Data,
      bodyFilePath: map["bodyFilePath"] as? String,
      responseBodyPath: map["responseBodyPath"] as? String,
      cancelTokenId: optionalUInt64Value(map["cancelTokenId"]),
      progressId: optionalUInt64Value(map["progressId"]),
      timeoutSeconds: optionalUInt64Value(map["timeoutSeconds"]),
      followRedirects: boolValue(map["followRedirects"], fallback: true)
    )
  }

  private func protocolMode(_ value: String?) -> VaneProtocolMode {
    switch value {
    case "http3ThenHttp2ThenHttp1":
      return .http3ThenHttp2ThenHttp1
    case "http2ThenHttp1":
      return .http2ThenHttp1
    case "http2Only":
      return .http2Only
    case "http1Only":
      return .http1Only
    default:
      return .http3Only
    }
  }

  private func stringMap(_ value: Any?) -> [String: String] {
    guard let map = value as? [String: Any] else {
      return [:]
    }
    return map.reduce(into: [String: String]()) { result, entry in
      result[entry.key] = "\(entry.value)"
    }
  }

  private func stringListMap(_ value: Any?) -> [String: [String]] {
    guard let map = value as? [String: Any] else {
      return [:]
    }
    return map.reduce(into: [String: [String]]()) { result, entry in
      result[entry.key] = stringList(entry.value)
    }
  }

  private func stringList(_ value: Any?) -> [String] {
    return (value as? [Any])?.map { "\($0)" } ?? []
  }

  private func boolValue(_ value: Any?, fallback: Bool) -> Bool {
    return value as? Bool ?? fallback
  }

  private func uint64Value(_ value: Any?, fallback: UInt64) -> UInt64 {
    guard let number = value as? NSNumber else {
      return fallback
    }
    return number.uint64Value
  }

  private func optionalUInt64Value(_ value: Any?) -> UInt64? {
    return (value as? NSNumber)?.uint64Value
  }

  private func requiredDictionary(_ value: Any?) throws -> [String: Any] {
    guard let dictionary = value as? [String: Any] else {
      throw VaneFlutterError.invalidArguments
    }
    return dictionary
  }

  private func requiredString(_ value: Any?) throws -> String {
    guard let string = value as? String else {
      throw VaneFlutterError.invalidArguments
    }
    return string
  }

  private func requiredInt64(_ value: Any?) throws -> Int64 {
    guard let number = value as? NSNumber else {
      throw VaneFlutterError.invalidArguments
    }
    return number.int64Value
  }
}

private enum VaneFlutterError: Error {
  case invalidArguments
}

private extension VaneResponse {
  func toMap() -> [String: Any] {
    return [
      "statusCode": Int(statusCode),
      "headers": headers,
      "body": FlutterStandardTypedData(bytes: body),
      "bodyFilePath": bodyFilePath as Any,
      "isSuccess": isSuccess,
      "url": url
    ]
  }
}

File: android/src/main/kotlin/com/inteniquetic/vane_flutter/VaneFlutterPlugin.kt
package com.inteniquetic.vane_flutter

import com.inteniquetic.vanekotlin.VaneClient
import com.inteniquetic.vanekotlin.VaneClientCertificate
import com.inteniquetic.vanekotlin.VaneClientConfig
import com.inteniquetic.vanekotlin.VaneProtocolMode
import com.inteniquetic.vanekotlin.VaneRequest
import com.inteniquetic.vanekotlin.VaneTlsVersion
import com.inteniquetic.vanekotlin.createDefaultConfig
import com.inteniquetic.vanekotlin.createVaneClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class VaneFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val nextHandle = AtomicLong(1)
    private val clients = ConcurrentHashMap<Long, VaneClient>()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "vane_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "createClient" -> createClient(call, result)
            "execute" -> execute(call, result)
            "closeClient" -> closeClient(call, result)
            "setCertificatePins" -> setCertificatePins(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        clients.values.forEach { it.destroy() }
        clients.clear()
        scope.cancel()
        channel.setMethodCallHandler(null)
    }

    private fun createClient(call: MethodCall, result: Result) {
        runCatching {
            val config = configFromMap(call.requiredMap("configuration"))
            val handle = nextHandle.getAndIncrement()
            clients[handle] = createVaneClient(config)
            result.success(handle)
        }.onFailure { error ->
            result.error("VANE_CREATE_CLIENT", error.message, null)
        }
    }

    private fun execute(call: MethodCall, result: Result) {
        val handle = call.requiredLong("handle")
        val requestMap = call.requiredMap("request")
        val client = clients[handle]
        if (client == null) {
            result.error("VANE_INVALID_CLIENT", "No Vane client exists for handle $handle", null)
            return
        }

        scope.launch {
            runCatching {
                client.executeRequest(requestFromMap(requestMap)).toMap()
            }.onSuccess { response ->
                result.success(response)
            }.onFailure { error ->
                result.error("VANE_EXECUTE", error.message, null)
            }
        }
    }

    private fun closeClient(call: MethodCall, result: Result) {
        val handle = call.requiredLong("handle")
        clients.remove(handle)?.destroy()
        result.success(null)
    }

    private fun setCertificatePins(call: MethodCall, result: Result) {
        val handle = call.requiredLong("handle")
        val host = call.requiredString("host")
        val pins = stringList(call.argument<Any?>("pins"))
        val client = clients[handle]
        if (client == null) {
            result.error("VANE_INVALID_CLIENT", "No Vane client exists for handle $handle", null)
            return
        }

        runCatching {
            client.setCertificatePins(host, pins)
            result.success(null)
        }.onFailure { error ->
            result.error("VANE_SET_CERTIFICATE_PINS", error.message, null)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun configFromMap(map: Map<String, Any?>): VaneClientConfig {
        return createDefaultConfig().apply {
            baseUrl = map["baseUrl"] as String?
            defaultHeaders = stringMap(map["defaultHeaders"])
            dnsOverrides = stringMap(map["dnsOverrides"])
            certificatePins = stringListMap(map["certificatePins"])
            cookiesEnabled = boolValue(map["cookiesEnabled"], cookiesEnabled)
            cookiePersistencePath = map["cookiePersistencePath"] as String?
            connectionPoolEnabled = boolValue(map["connectionPoolEnabled"], connectionPoolEnabled)
            maxIdleConnections = ulongValue(map["maxIdleConnections"], maxIdleConnections)
            connectionIdleTimeoutSeconds = ulongValue(
                map["connectionIdleTimeoutSeconds"],
                connectionIdleTimeoutSeconds
            )
            retryMaxAttempts = ulongValue(map["retryMaxAttempts"], retryMaxAttempts)
            retryInitialDelayMillis = ulongValue(
                map["retryInitialDelayMillis"],
                retryInitialDelayMillis
            )
            retryMaxDelayMillis = ulongValue(map["retryMaxDelayMillis"], retryMaxDelayMillis)
            retryUnsafeMethods = boolValue(map["retryUnsafeMethods"], retryUnsafeMethods)
            maxRequestBodyBytes = ulongValue(map["maxRequestBodyBytes"], maxRequestBodyBytes)
            maxResponseBodyBytes = ulongValue(map["maxResponseBodyBytes"], maxResponseBodyBytes)
            timeoutSeconds = nullableULongValue(map["timeoutSeconds"])
            followRedirects = boolValue(map["followRedirects"], followRedirects)
            userAgent = map["userAgent"] as String?
            protocolMode = protocolMode(map["protocolMode"] as String?)
            proxyUrl = map["proxyUrl"] as String?
            proxyAuthorization = map["proxyAuthorization"] as String?
            maxRedirects = uintValue(map["maxRedirects"], maxRedirects)
            tlsMinVersion = tlsVersion(map["tlsMinVersion"] as String?)
            tlsMaxVersion = tlsVersion(map["tlsMaxVersion"] as String?)
            customRootCertificates = stringList(map["customRootCertificates"])
            clientCertificate = clientCertificate(map["clientCertificate"])
        }
    }

    private fun requestFromMap(map: Map<String, Any?>): VaneRequest {
        return VaneRequest(
            url = map["url"] as String,
            method = (map["method"] as String?) ?: "GET",
            headers = stringMap(map["headers"]),
            queryParams = stringMap(map["queryParams"]),
            body = map["body"] as ByteArray?,
            bodyFilePath = map["bodyFilePath"] as String?,
            responseBodyPath = map["responseBodyPath"] as String?,
            cancelTokenId = ulongValue(map["cancelTokenId"], 0uL).takeIf { it != 0uL },
            progressId = ulongValue(map["progressId"], 0uL).takeIf { it != 0uL },
            timeoutSeconds = nullableULongValue(map["timeoutSeconds"]),
            followRedirects = boolValue(map["followRedirects"], true)
        )
    }

    private fun com.inteniquetic.vanekotlin.VaneResponse.toMap(): Map<String, Any?> {
        return mapOf(
            "statusCode" to statusCode.toInt(),
            // Ordered [name, value] pairs, duplicates preserved, set-cookie
            // inline in arrival position — the same shape the FFI path reads.
            // The Dart parser and both plugins move together, or this path
            // silently mis-parses.
            "headers" to headers.map { listOf(it.name, it.value) },
            "body" to body,
            "bodyFilePath" to bodyFilePath,
            "isSuccess" to isSuccess,
            "url" to url,
            "httpVersion" to httpVersion?.name?.lowercase(),
            // Kept at parity with the FFI path: a field present on one Flutter
            // transport and absent on the other only reproduces on some setups.
            "remoteIp" to remoteIp
        )
    }

    private fun tlsVersion(value: String?): VaneTlsVersion? {
        return when (value) {
            "tls12" -> VaneTlsVersion.TLS12
            "tls13" -> VaneTlsVersion.TLS13
            else -> null
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun clientCertificate(value: Any?): VaneClientCertificate? {
        val map = value as? Map<Any?, Any?> ?: return null
        val certificatePem = map["certificatePem"] as? String ?: return null
        val privateKeyPem = map["privateKeyPem"] as? String ?: return null
        return VaneClientCertificate(certificatePem, privateKeyPem)
    }

    private fun protocolMode(value: String?): VaneProtocolMode {
        return when (value) {
            "http3ThenHttp2ThenHttp1" -> VaneProtocolMode.HTTP3_THEN_HTTP2_THEN_HTTP1
            "http2ThenHttp1" -> VaneProtocolMode.HTTP2_THEN_HTTP1
            "http2Only" -> VaneProtocolMode.HTTP2_ONLY
            "http1Only" -> VaneProtocolMode.HTTP1_ONLY
            else -> VaneProtocolMode.HTTP3_ONLY
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun stringMap(value: Any?): Map<String, String> {
        return (value as? Map<Any?, Any?>)
            ?.mapKeys { it.key.toString() }
            ?.mapValues { it.value.toString() }
            ?: emptyMap()
    }

    @Suppress("UNCHECKED_CAST")
    private fun stringListMap(value: Any?): Map<String, List<String>> {
        return (value as? Map<Any?, Any?>)
            ?.mapKeys { it.key.toString() }
            ?.mapValues { entry ->
                stringList(entry.value)
            }
            ?: emptyMap()
    }

    private fun stringList(value: Any?): List<String> {
        return (value as? List<Any?>)?.map { it.toString() } ?: emptyList()
    }

    private fun boolValue(value: Any?, fallback: Boolean): Boolean {
        return value as? Boolean ?: fallback
    }

    private fun ulongValue(value: Any?, fallback: ULong): ULong {
        return (value as? Number)?.toLong()?.toULong() ?: fallback
    }

    private fun uintValue(value: Any?, fallback: UInt): UInt {
        return (value as? Number)?.toInt()?.toUInt() ?: fallback
    }

    private fun nullableULongValue(value: Any?): ULong? {
        return (value as? Number)?.toLong()?.toULong()
    }

    @Suppress("UNCHECKED_CAST")
    private fun MethodCall.requiredMap(key: String): Map<String, Any?> {
        return argument<Map<String, Any?>>(key)
            ?: throw IllegalArgumentException("Missing $key")
    }

    private fun MethodCall.requiredLong(key: String): Long {
        return when (val value = argument<Any>(key)) {
            is Number -> value.toLong()
            else -> throw IllegalArgumentException("Missing $key")
        }
    }

    private fun MethodCall.requiredString(key: String): String {
        return argument<String>(key) ?: throw IllegalArgumentException("Missing $key")
    }
}
